R/OptimizerChain.R

```diff
@@ -0,0 +1,163 @@
+#'
+#'codomain = ps(y = p_dbl(tags = "minimize"))
+#'
+#'objective_function = function(xs) {
+#'  list(y = as.numeric(xs)^2)
+#'}
+#'
+#'objective = ObjectiveRFun$new(
+#'  fun = objective_function,
+#'  domain = domain,
+#'  codomain = codomain
+#')
+#'
+#'terminator = trm("evals", n_evals = 10)
+#'
+#'instance = OptimInstanceSingleCrit$new(
+#' objective = objective,
+#' search_space = search_space,
+#' terminator = terminator
+#')
+#'
+#'optimizer = opt("chain",
+#'  optimizers = list(opt("random_search"), opt("grid_search")),
+#'  terminators = list(trm("evals", n_evals = 5), trm("evals", n_evals = 5))
+#')
+#'
+#'optimizer$optimize(instance)
+#'
+#' # random restarts
+#'instance$archive$clear()  # reset instance
+#'instance$terminator = trm("none")
+#'optimizer = opt("chain",
+#'  optimizers = list(opt("gensa"), opt("gensa")),
+#'  terminators = list(trm("evals", n_evals = 10), trm("evals", n_evals = 10))
+#')
+#'optimizer$optimize(instance)
+OptimizerChain = R6Class("OptimizerChain", inherit = Optimizer,
+  public = list(
+
+    #' @description
+    #' Creates a new instance of this [R6][R6::R6Class] class.
+    #'
+    #' @param optimizers (list of [Optimizer]s).
+    #' @param terminators (list of [Terminator]s | NULL).
+    initialize = function(optimizers, terminators = rep(list(NULL), length(optimizers))) {
+      assert_list(optimizers, types = "Optimizer", any.missing = FALSE)
+      assert_list(terminators, types = c("Terminator", "NULL"), len = length(optimizers))
+
+      param_sets = vector(mode = "list", length = length(optimizers))
+      ids_taken = character(0L)
+      # for each optimizer check whether the id of the param_set
+      # (decuded from the optimizer class) is already taken;
+      # if necessary postfix the id
+      for (i_opt in seq_along(optimizers)) {
+        opt = optimizers[[i_opt]]
+        ps = opt$param_set$clone(deep = TRUE)
+        ps$set_id = class(opt)[[1L]]
+        try_postfix = 0L
+        while (ps$set_id %in% ids_taken) {
+          try_postfix = try_postfix + 1L
+          ps$set_id = paste0(class(opt)[[1L]], "_", try_postfix)
+        }
+        ids_taken[[i_opt]] = ps$set_id
+        param_sets[[i_opt]] = ps
+      }
+      private$.ids = map_chr(param_sets, "set_id")
+      super$initialize(
+        param_set = ParamSetCollection$new(param_sets),
+        param_classes = Reduce(intersect, mlr3misc::map(optimizers, "param_classes")),
+        properties = Reduce(intersect, mlr3misc::map(optimizers, "properties")),
+        packages = unique(unlist(mlr3misc::map(optimizers, "packages")))
+      )
+      private$.optimizers = optimizers
+      private$.terminators = terminators
+    }
+  ),
+
+  private = list(
+    .optimizers = NULL,
+    .terminators = NULL,
+    .ids = NULL,
+
+    .optimize = function(inst) {
+      terminator = inst$terminator
+      on.exit({inst$terminator = terminator})
+      inner_inst = inst$clone(deep = TRUE)
```

Comment:
would be good if this could be avoided, so when an error happens in the inner optimization then the user's `inst` at least has the partial optimization progress. Did I do this in my original implementation?

Reply:
We specifically do the `on.exit()` thing so we can modify the `inst`'s `$terminator` temporarily without needing to clone, or did something get uncovered that does make this necessary?